[4.0] Pagination seo friendly

rajkumaar23 opened the pull request:
Pull Request for Issue #26881 

### Summary of Changes
- Appends a `limitstart` param to the action on pagination links


### Testing Instructions
- Ensure the pagination doesn't break with this PR.
- Verify if `limitstart` param is appended & modified while paginating.
- Check if proper page is rendered when `limitstart` is modified in address bar.


### Expected result



### Actual result



### Documentation Changes Required
NIL

rajkumaar23 committed:
Removes unnecessary limit calculation in pagination

rajkumaar23 committed:
Merge branch '4.0-dev' into 4.0-dev

Commit:
Merge branch '4.0-dev' into 4.0-dev

rajkumaar23 committed:
Merge remote-tracking branch 'joomla/4.0-dev' into 4.0-dev

Quy commented:
Please revert coding style changes per [Joomla! Coding Standards](https://developer.joomla.org/coding-standards/html.html).

rajkumaar23 commented:
> Please revert coding style changes per [Joomla! Coding Standards](https://developer.joomla.org/coding-standards/html.html).

Okay sure, Can you let me know how to populate the pages that use pagination? Because I don’t seem to have enough data to paginate.

Quy commented:
Try installing `Testing Sample Data'. Then do a search on the frontend.

rajkumaar23 commented:
> Try installing `Testing Sample Data`. Then do a search on the frontend.

Thanks @Quy ! I was able tot test with the dummy data and I've reverted as per coding standards and fixed another bug in the PR too.

rajkumaar23 commented:
Thanks @Quy ! I was able to test with the dummy data and I've reverted as per coding standards and fixed another bug in the PR too.

rajkumaar23 commented:
@Quy Could you please review this PR ?

ChristineWk commented:
I have tested this item :red_circle: unsuccessfully on 81a7296c146a10982e01d857470335ea61c2771b<br /><br />When going to i.e. page 2 (on pagination link) I get error 404.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27489">issues.joomla.org/tracker/joomla-cms/27489</a>.</sub>

rajkumaar23 commented:
@ChristineWk Can you please check now ?

ChristineWk commented:
I have tested this item :red_circle: unsuccessfully on e3a52e5514ca7dba479c17a653905932fee700af<br /><br />Error 404 has gone. On 2nd page > in address bar I see: blog?limitstart=5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27489">issues.joomla.org/tracker/joomla-cms/27489</a>.</sub>

ChristineWk commented:
I have tested this item :white_check_mark: successfully on e3a52e5514ca7dba479c17a653905932fee700af<br /><br />Error 404 has gone. On 2nd page > in address bar I see: blog?limitstart=5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27489">issues.joomla.org/tracker/joomla-cms/27489</a>.</sub>

Quy commented on layouts/joomla/pagination/link.php:
```diff
@@ -54,6 +54,17 @@
 {
+	$uriUpdatingScript = "var re = new RegExp('([?&])' + 'limitstart' + '=.*?(&|$)', 'i');
+						  var uri = document.adminForm.action;
+    	   				  var separator = uri.indexOf('?') !== -1 ? '&' : '?';
```
Please use tabs and not spaces for indentation.

rajkumaar23 replied:
Okay, sure.

rajkumaar23 replied:
@Quy I have updated.

Hackwar commented:
I agree that the initial issue is a bug, but I don't understand why the method from J3 isn't still used. Can someone please explain why we should do it this way?

Fedik commented:
I am sorry to say, but the fix are wrong, such JavaScript hacking not acceptable

rajkumaar23 commented:
> I am sorry to say, but the fix are wrong, such JavaScript hacking not acceptable

Since the pagination itself is done via Javascript, I couldn’t think of any other way to update the URL with the limitstart value. Could you suggest a better idea?

rajkumaar23 commented:
> I agree that the initial issue is a bug, but I don't understand why the method from J3 isn't still used. Can someone please explain why we should do it this way?

Can you please let me know how it was done in J3 ? I’m new to Joomla! and I checked the branches here, and everywhere it’s done in JavaScript only.

rajkumaar23 committed:
[4.0] Tries to make pagination seo friendly

rajkumaar23 committed:
Merge branch '4.0-dev' into pagination-seo-friendly

SharkyKZ commented:
In J3 frontend template had an override without JS in it.

Maybe in 4.0 we could do the other way around, make it JS-free by default and add an override for backend template instead.

rajkumaar23 commented:
I guess then the complete pagination code has to change if it has to be no-JS. All I’ve done is append the limitstart param to the form action, added to the already existing JS code that modifies the hidden ‘limitstart’ field in the form and triggers it. As far as this issue is concerned, the task was to make different URLs render different content.

Fedik commented:
You put all your "multi-line" javascript code in to oneclick attribute, do not do this ever.
Also please check how the `form` element works, you no need javascript to append a value that already there (use appropriate submit method GET/POST)

About correct fix, the pagination need to be a list of links, for frontend.
That @SharkyKZ already said

rajkumaar23 commented:
@Fedik I'm sorry, but I couldn't understand what you're trying to say. If I'm not wrong, do you mean all this pagination forms has to be GET, so the URL will have `limitstart` appended automatically ?

Fedik commented:
Pagination should be a list of links, without javascript, without form.

rajkumaar23 commented:
> Pagination should be a list of links, without javascript, without form.

@Fedik Okay, so you mean to say the current implementation for pagination has to be changed to links. I thought that was out of scope for the related issue. I'll check how to do it.

Fedik commented:
Check @SharkyKZ comment, it should be changed only for the site, and stay unchanged for administrator.

rajkumaar23 commented:
> Check @SharkyKZ comment, it should be changed only for the site, and stay unchanged for administrator.

Sorry! I'm new to Joomla! Can I get some guidance on what's the `override` he was talking about ?

Fedik commented:
I think these 2 links

https://docs.joomla.org/Layout_Overrides_in_Joomla#JLayout_Micro-Layout_Overrides
https://docs.joomla.org/J3.x:Sharing_layouts_across_views_or_extensions_with_JLayout

SharkyKZ commented:
Templates support layout overrides. You can copy `layouts/joomla/pagination/link.php` to `templates/cassiopeia/html/layouts/joomla/pagination/link.php` to create an override for frontend template and remove JS code there.

Or the other way around, copy the file as-is to backend template: `administrator/templates/atum/html/layouts/joomla/pagination/link.php` and then edit original file to remove JS code.

There are other alternatives to consider like creating multiple standard layouts and making them configurable in pagination calls.

rajkumaar23 committed:
Removes unnecessary JS and overrides pagination layout in frontend

rajkumaar23 committed:
Merge remote-tracking branch 'origin/pagination-seo-friendly' into pagination-seo-friendly

rajkumaar23 commented:
@SharkyKZ Can you please check my latest commit if I've correctly understood what you told ?

zero-24 commented on templates/cassiopeia/html/layouts/joomla/pagination/link.php:
```diff
@@ -0,0 +1,107 @@
+}
+
+// Retrieving the URL for anchor with limitstart param added/modified
+$currentURL = "http://$_SERVER[HTTP_HOST]$_SERVER[REQUEST_URI]";
```
hmm we should not hardcode http here we might should also better use `JUri` / `\Joomla\CMS\Uri\Uri`

rajkumaar23 replied:
Sure, thanks.

rajkumaar23 replied:
Is it fine now?

rajkumaar23 committed:
Imports JUri

Fedik commented on templates/cassiopeia/html/layouts/joomla/pagination/link.php:
```diff
@@ -0,0 +1,107 @@
+
+$params['limitstart'] = $item->base ?: 0;
+$urlParts['query'] = http_build_query($params);
+$updatedURL = http_build_url($urlParts);
```
why you doing this?

rajkumaar23 replied:
That is the href I've set for the pagination button. How else would I navigate to the next `page` without `limitstart` ?

Fedik replied:
use `href="<?php echo $item->link; ?>"` it already contain all you need

rajkumaar23 replied:
I did var_dump on `$item->link` when I was on `/article-category-list` page, but the value was not current URL.

Fedik replied:
what URL was in your var_dump?
the `$item->link` contain a correct link for every pagination item, you no need to modify it, it already prepared in `Pagination` class

rajkumaar23 replied:
I think I did a var_dump($item) in administrator page, and received different `link`. Thanks, will use that here.

rajkumaar23 replied:
Sorry, that was my mistake. I have used it now, thanks :)

rajkumaar23 committed:
Removes unnecessary computation of link

rajkumaar23 commented:
@Fedik I think now everything is sorted.

Fedik commented on templates/cassiopeia/html/layouts/joomla/pagination/link.php:
```diff
@@ -0,0 +1,90 @@
+if ($displayData['active'])
+{
+
+	if ($item->base > 0)
```
you can remove this section, it was need for javascript

Fedik commented on templates/cassiopeia/html/layouts/joomla/pagination/link.php:
```diff
@@ -0,0 +1,90 @@
+	}
+
+	$class = 'active';
+	$onClick = 'document.adminForm.' . $item->prefix . $limit . '; Joomla.submitform();return false;';
```
and this also not need anymore

rajkumaar23 replied:
Done :+1: Removed.

rajkumaar23 committed:
Removes unnecessary JS

Fedik commented:
Now it looks much better than first version of it :smiley: 
Left only a testing, I guess.

rajkumaar23 commented:
> Now it looks much better than first version of it smiley
> Left only a testing, I guess.

Haha. This is my first PR with so many comments. OpenSource contributions are indeed best in learning. Thanks! 🔥️

Quy commented on templates/cassiopeia/html/layouts/joomla/pagination/link.php:
```diff
@@ -0,0 +1,79 @@
+		break;
+
+	// Check for "Prev" item
+	case $item->text === Text::_('JPREV') :
```
```suggestion
	case Text::_('JPREV') :
```

rajkumaar23 replied:
Done.

rajkumaar23 committed:
Update templates/cassiopeia/html/layouts/joomla/pagination/link.php

Co-Authored-By: Quy <quy@fluxbb.org>

brianteeman commented:
I am coming to this late but to me it is crazy that we need to create an override. If we can't make the core work with the default templates that it ships with there is something very wrong

rajkumaar23 commented:
> I am coming to this late but to me it is crazy that we need to create an override. If we can't make the core work with the default templates that it ships with there is something very wrong
Alright then, does having the start param in URL on administrator routes cause any issue? If not, I don’t think this override is required. I can edit the original link.php, but as I was told in this thread that only frontend needs to have that param in the URL, override was the only possible way

brianteeman commented:
Why does it have to be the same layout for both site and admin?

rajkumaar23 commented:
I didn’t get you @brianteeman :(

Fedik commented:
@brianteeman there was comment @SharkyKZ suggested to make "this" as default, and then make override for admin template (should be simple), or make them configurable in pagination calls to use different layout depend from needs (this more advanced topic).

Admin use different pagination to keep adminForm "state". The form has extra fields, like CSR token, option, view etc, and can have more depend from component. As far as I know.

brianteeman commented:
My view is that it should be
> or make them configurable in pagination calls to use different layout depend from needs

rajkumaar23 commented:
@Fedik What do you think could I possibly do to fix this ?

Fedik commented:
It is good as it is for now, it fixes the issue.

What was suggested is much complex work to do, and can be done any time later.

rajkumaar23 commented:
> It is good as it is for now, it fixes the issue.
> 
> What was suggested is much complex work to do, and can be done any time later.

Can someone test this ?

brianteeman commented:
Unless I am very much mistaken it is not complex at all
See #27516 for a very quick example

Fedik commented:
I think we have a different look on what is "configurable pagination" :wink: 

With your suggestion it still not possible to use simple links in backend, or javascript version on front end. It hardly can call "configurable".

The Pagination class allow to specify the layout, that allow to have 2 (an more) different layout at same time
https://github.com/joomla/joomla-cms/blob/4779815c0346c1ba273a6cac4b6d346771f1c8ba/libraries/src/Pagination/Pagination.php#L436-L442

A complexity here is to check all call of `$pagination->getPaginationLinks()` (both admin/site side) and adjust it to use specific layout, and make sure all works.

brianteeman commented:
> I think we have a different look on what is "configurable pagination" 😉

Maybe so but my PR resolves the issue in the original report without creating template overrides

Fedik commented:
yeap, that also ok, I don't mind

brianteeman commented:
as i said before 
> I am coming to this late but to me it is crazy that we need to create an override. If we can't make the core work with the default templates that it ships with there is something very wrong
